### PR TITLE
feat(substrate-aio): add ws-port argument

### PR DESCRIPTION
**Pull request by RafaelAPB**
This PR allows developers to choose the port where the saio exposes.

However, there is a bug: the substrate connector cannot connect to the saio running on the custom port. Perhaps it is an issue with port exposing?

### CODE REVIEW

**Comment by petermetz** [comment]
> However, there is a bug: the substrate connector cannot connect to the saio running on the custom port. Perhaps it is an issue with port exposing?

@RafaelAPB Is the above still an issue?

**Comment by RafaelAPB** [comment]
@petermetz yes - I believe it may have something to do with the fact that the ledger is run via a script rather than CMD.

**Comment by petermetz** [comment]
> @petermetz yes - I believe it may have something to do with the fact that the ledger is run via a script rather than CMD.

@RafaelAPB Is there a test case/set of instructions that I can run locally on my machine to reproduce the issue you are having?

**Comment by RafaelAPB** [comment]
@petermetz it would be creating the SAIO instance and then running a Polkadot connector with the provider set as the WS provider that the SAIO exposes. I will send more specific instructions later

**Comment by petermetz** [comment]
@RafaelAPB Thank you, that will be very helpful, I'm super short on time ;-(

**Comment by RafaelAPB** [comment]
@petermetz  sorry for the looong time to do this, but I think this reproduces the issue.
The goal is for the test ledger to be able to be ran in parallel in several test files (one creates a test ledger listening to connections let's say on port 9944, other in 9946, and so on).

To showcase that the previous commit works: on a terminal, run:

`docker  run --expose 9944 --env PORT=9944 --env WS_PORT=9944 --env DOCKER_PORT=9944 --env WORKING_DIR=/var/www/node-template --env CONTAINER_NAME=contracts-node-template-ovl --env CARGO_HOME=/var/www/node-template/.cargo  -p 9944:9944  ghcr.io/hyperledger/cactus-substrate-all-in-one:2021-10-15---feat-1274` 

It now has started listening to connections on pot 9944. We want to run a SAIO concurrently such that it will listen on a different port. I've tried:

`docker  run --expose 9946 --env PORT=9946 --env WS_PORT=9946 --env DOCKER_PORT=9946 --env WORKING_DIR=/var/www/node-template --env CONTAINER_NAME=contracts-node-template-ovl --env CARGO_HOME=/var/www/node-template/.cargo  -p 9946:9946  ghcr.io/hyperledger/cactus-substrate-all-in-one:2021-10-15---feat-1274`

but it still listens to connections on the 9944 port. Everything works, as Substrate clients can access the test ledger. The connector @CatarinaPedreira developed, PR #1490  should in principle help us debug this. Now, consider the image I've created based on the commits from this PR: rafaelapb/cactus-substrate-aio:2021-11-02-fix
Run

`docker  run --expose 9944 --env PORT=9944 --env WS_PORT=9944 --env DOCKER_PORT=9944 --env WORKING_DIR=/var/www/node-template --env CONTAINER_NAME=contracts-node-template-ovl --env CARGO_HOME=/var/www/node-template/.cargo  -p 9944:9944  rafaelapb/cactus-substrate-aio:2021-11-02-fix`

and 

` docker  run --expose 9946 --env PORT=9646 --env WS_PORT=9946 --env DOCKER_PORT=9946 --env WORKING_DIR=/var/www/node-template --env CONTAINER_NAME=contracts-node-template-ovl --env CARGO_HOME=/var/www/node-template/.cargo  -p 9946:9946  ghcr.io/hyperledger/cactus-substrate-all-in-one:2021-10-15---feat-1274`

Now the test ledgers are listening to connections on 9944 and 9946, respectively - however, a Substrate client cannot connect to the test ledger. If you send a sigterm to a container instantiated with this last image, it won't stop the test ledger (you need to manually kill the process), whereas with the other image it receives sigterms. I think this has to do with the fact we are running a bespoke script as the entrypoint to the docker container; this inability to receive sigterms is probably also causing the clients no to be able to connect.

**Comment by petermetz** [comment]
@RafaelAPB Thank you for the specific commands. Unfortunately I couldn't reproduce the issue because the third command crashes the container with the logs pasted below. (which then made it impossible for me to verify if it would listen on the correct port or not).
Could you please double check if the `rafaelapb/cactus-substrate-aio:2021-11-02-fix` image version is working for you with the command you put in your previous message?

> If you send a sigterm to a container instantiated with this last image, it won't stop the test ledger (you need to manually kill the process), whereas with the other image it receives sigterms. I think this has to do with the fact we are running a bespoke script as the entrypoint to the docker container;

I agree, we might be better off with `tini` or similar running the bash script, but for now let's just focus on getting to a point where we can both reproduce the issue without problems.

> this inability to receive sigterms is probably also causing the clients no to be able to connect.

I'm not so sure about this part though. Based on my testing the newer image (third command from your previous message) just doesn't work and that's why the connection is seemingly refused as well. I'm not a very big expert on operating systems, but I'd assume that signal handlers don't modify the behavior of network sockets listening on certain ports (but then again, stranger things have happened so I want to continue to investigate this with you)

```sh
$ docker run --expose 9944 --env PORT=9944 --env WS_PORT=9944 --env DOCKER_PORT=9944 --env WORKING_DIR=/var/www/node-template --env CONTAINER_NAME=contracts-node-template-ovl --env CARGO_HOME=/var/www/node-template/.cargo -p 9944:9944 rafaelapb/cactus-substrate-aio:2021-11-02-fix
Unable to find image 'rafaelapb/cactus-substrate-aio:2021-11-02-fix' locally
2021-11-02-fix: Pulling from rafaelapb/cactus-substrate-aio
b380bbd43752: Already exists 
3f95576566d2: Pull complete 
63afad639ee4: Pull complete 
d9591b534643: Pull complete 
932ec488e402: Pull complete 
badb25e4399e: Pull complete 
9c97448206cd: Pull complete 
2190e9953833: Pull complete 
54f44c3ca213: Pull complete 
947ce6feea9d: Pull complete 
Digest: sha256:799d794699f0b835a6da10101d7e597addf804fb73a3303dfd85de94c42b322c
Status: Downloaded newer image for rafaelapb/cactus-substrate-aio:2021-11-02-fix
+ echo ENV:PORT=9944
+ echo ENV:WS_PORT=9944
ENV:PORT=9944
ENV:WS_PORT=9944
+ /var/www/node-template/.cargo/bin/substrate-contracts-node --dev --port 9944 --ws-port 9944
2021-12-13 22:43:39 Running in --dev mode, RPC CORS has been disabled.    
2021-12-13 22:43:39 Substrate Contracts Node    
2021-12-13 22:43:39 ✌️  version 0.1.0-3f63edf-x86_64-linux-gnu    
2021-12-13 22:43:39 ❤️  by Parity Technologies <admin@parity.io>, 2021-2021    
2021-12-13 22:43:39 📋 Chain specification: Development    
2021-12-13 22:43:39 🏷 Node name: sour-shoes-6214    
2021-12-13 22:43:39 👤 Role: AUTHORITY    
2021-12-13 22:43:39 💾 Database: RocksDb at /root/.local/share/substrate-contracts-node/chains/dev/db/full    
2021-12-13 22:43:39 ⛓  Native runtime: substrate-contracts-node-100 (substrate-contracts-node-1.tx1.au1)    
2021-12-13 22:43:39 🔨 Initializing Genesis block/state (state: 0x0a8d…2d8a, header-hash: 0x79c9…5ee4)    
2021-12-13 22:43:39 👴 Loading GRANDPA authority set from genesis on what appears to be first startup.    
2021-12-13 22:43:40 ⏱  Loaded block-time = 6s from block 0x79c9523d97c2cfc26ed8a8184e53c970e2bdebe79860c9f9019aaba6ee965ee4    
2021-12-13 22:43:40 🏷 Local node identity is: 12D3KooWKM7MiDgAof7RzAeSvx9xH3keR4aLG6NL3KqbzJmECtnG    
2021-12-13 22:43:40 📦 Highest known block at #0    
2021-12-13 22:43:40 〽️ Prometheus exporter started at 127.0.0.1:9615    
2021-12-13 22:43:40 Unable to listen on 127.0.0.1:9944    

====================

Version: 0.1.0-3f63edf-x86_64-linux-gnu

   0: sp_panic_handler::set::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/panicking.rs:628:17
   2: std::panicking::begin_panic::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: std::panicking::begin_panic
   5: tokio::runtime::blocking::shutdown::Receiver::wait
   6: tokio::runtime::blocking::pool::BlockingPool::shutdown
   7: <tokio::runtime::blocking::pool::BlockingPool as core::ops::drop::Drop>::drop
   8: core::ptr::drop_in_place<tokio::runtime::blocking::pool::BlockingPool>
   9: jsonrpc_ws_server::server::Server::start
  10: sc_rpc_server::inner::start_ws
  11: sc_service::start_rpc_servers::{{closure}}
  12: sc_service::start_rpc_servers
  13: sc_service::builder::spawn_tasks
  14: substrate_contracts_node::service::new_full
  15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  16: tokio::park::thread::CachedParkThread::block_on
  17: tokio::runtime::thread_pool::ThreadPool::block_on
  18: tokio::runtime::Runtime::block_on
  19: sc_cli::runner::Runner<C>::run_node_until_exit
  20: substrate_contracts_node::command::run
  21: substrate_contracts_node::main
  22: std::sys_common::backtrace::__rust_begin_short_backtrace
  23: std::rt::lang_start::{{closure}}
  24: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/core/src/ops/function.rs:259:13
      std::panicking::try::do_call
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/panicking.rs:403:40
      std::panicking::try
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/panicking.rs:367:19
      std::panic::catch_unwind
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/panic.rs:129:14
      std::rt::lang_start_internal::{{closure}}
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/rt.rs:45:48
      std::panicking::try::do_call
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/panicking.rs:403:40
      std::panicking::try
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/panicking.rs:367:19
      std::panic::catch_unwind
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/panic.rs:129:14
      std::rt::lang_start_internal
             at rustc/51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8/library/std/src/rt.rs:45:20
  25: main
  26: __libc_start_main
  27: _start


Thread 'main' panicked at 'Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context.', /var/www/node-template/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.11.0/src/runtime/blocking/shutdown.rs:51

This is a bug. Please report it at:

        https://github.com/paritytech/canvas-node/issues/new

```

**Comment by RafaelAPB** [comment]
@petermetz could you exchange the third command for:
`docker run --expose 9948   --env WS_PORT=9948 --env DOCKER_PORT=9948 --env WORKING_DIR=/var/www/node-template --env CONTAINER_NAME=contracts-node-template-ovl --env CARGO_HOME=/var/www/node-template/.cargo -p 9948:9948 rafaelapb/cactus-substrate-aio:2021-11-02-fix`

Effectively changes port 9944 for 9948. Seems to work this way

**Comment by petermetz** [comment]
@RafaelAPB Thanks, now the third command works too with the adjustment. So now I have the 2 containers running (command 3 and 4) and want to reproduce this part of what you said:

> however, a Substrate client cannot connect to the test ledger.

Is there a test case for that that I can run to reproduce? Or how else to check it?

The sigterm problem I already verified on my side as well as an issue and I think it's because of the switch from the exec form to the shell form for your CMD declaration in the dockerfile. More details are here: https://hynek.me/articles/docker-signals/
TLDR: I'd later try switching it back and see if we went full circle and arrived back at the env vars not being effective problem, BUT before I'd get to any of that we first just need to figure out the port listening issues.

**Comment by RafaelAPB** [comment]
@petermetz sorry for the delay. As the Polkadot connector development advances, I will be able to test this issue with our connector. I'll let you know if there are developments

**Comment by RafaelAPB** [comment]
@petermetz i've solved this issue. The test ledger can now be ran in different ports. I've pushed an image, afaelapb/cactus-substrate-aio:2022-03-28; can we publish this to the official Hyperledger repo?

**Review by takeutak — APPROVED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
> @petermetz i've solved this issue. The test ledger can now be ran in different ports. I've pushed an image, rafaelapb/cactus-substrate-aio:2022-03-28; can we publish this to the official Hyperledger repo? 

@RafaelAPB I published the image from this PR as `ghcr.io/hyperledger/cactus-substrate-all-in-one:2022-03-29--1496` just now. Please update the code to use that as the new default and then it's good to go from my side. Thank you again for the contributions! :-)

**Comment by RafaelAPB** [comment]
Thanks for all the help Peter, wouldn't be possible without you!

 I've also made some changes to ``substrate-test-ledger.ts``. I suppose those changes will be available in the npm packages in the next release? Is the alternative one to publish themselves the package and start using it right away?

**Comment by RafaelAPB** [comment]
@petermetz I think this is ready for review!

**Review by petermetz — APPROVED** [review]
@RafaelAPB LGTM, thank you!

**Review by petermetz — CHANGES_REQUESTED** [review]
@RafaelAPB  My bad, could you please squash the commits into one?

**Review by petermetz — APPROVED** [review]
@RafaelAPB Thank you, LGTM!
